### DISCUSSION TITLE
[4] Dinder - Class diagrams

### DISCUSSION THREAD

**Discussion by Manuel-F-04** [dhbw-ka-tinf22b5-dinder/Dinder#9] [Projektblog]
### Class diagram - frontend:
[picture - class diagram frontend](https://github.com/dhbw-ka-tinf22b5-dinder/Dinder-SRS/blob/main/Diagramme/Klassendiagramme/Frontend.png)

**Description:**
Here's a brief description of the classes and their relationships:

LoginPage: This class represents the page where users can log in. It includes text fields for the username and password, along with a login button.

RegistrationPage: This class represents the page where users can register. It includes text fields for the username, password, password confirmation, and email address.

MainPage: This class represents the main page of the application. It illustrates the typical Tinder system, allowing the contractor to swipe through advertisements.

ChatWindow: This class represents the chat window where users can exchange messages. It includes fields for chat messages and the input field for new messages.

ListedAdvertisements: This class represents a list of listed advertisements. It includes fields for listed ads.

AcceptedAdvertisements: This class represents a list of accepted advertisements. It also includes fields for listed ads.

NavigationBar: This class represents the navigation bar displayed on various pages. It includes buttons to switch between the "Your Chats," "My Listed Advertisements," "User," and "Main Page."

SingleAdvertisement: This class represents a single ad. It includes fields for the ad image, employer name, ad title, date, time, and notes.

### Class diagram - backend:
[picture - class diagram backend](https://github.com/dhbw-ka-tinf22b5-dinder/Dinder-SRS/blob/main/Diagramme/Klassendiagramme/Backend.png)

**Description:**
A user can both create an advert and apply for others.
Due to the cardinalities, an intermediate table is required for the second option.
Several chats can be assigned to an advertisement, whereby each chat belongs to an advertisement.
A chat consists of any number of messages that are assigned to a specific chat.
A chat is also assigned to two users (contributor, employer), whereby a user can use any number of chats.

The application starts in the DinderApplication class, which then forwards the start call to the spring framework.
This in turn creates all needed instances of the classes in the application.
The repository-classes contain default methods to modify the entities in the database and to find them by an optional criteria.
The Service holds instances of all the repository-classes, it exposes their functionality to all other classes.
It also provides additional functionality that require some logic to be implemented.
The DinderRestController defines the HTTP-endpoints for the application, which allow outside access to see and modify the data.

#### 1 Form page for a new advertisement / edit an existing advertisement
There is the possibility to fill out a form page to create a new service advertisement.

Sequence Diagram:
[form page_Sequence](https://github.com/dhbw-ka-tinf22b5-dinder/Dinder-SRS/tree/main/Diagramme/Sequenzdiagramme/FormPageAdvertisement.png)

Changed "User" to "Employer", Added lifelines

#### 2 Change - Lifelines
The components in each diagram now have lifelines.

Sequence diagrams:

[chat_Sequence](https://github.com/dhbw-ka-tinf22b5-dinder/Dinder-SRS/tree/main/Diagramme/Sequenzdiagramme/ChatDinder.png)

[login page_Sequence](https://github.com/dhbw-ka-tinf22b5-dinder/Dinder-SRS/tree/main/Diagramme/Sequenzdiagramme/loginPageSource.png)

[registration page_Sequence](https://github.com/dhbw-ka-tinf22b5-dinder/Dinder-SRS/tree/main/Diagramme/Sequenzdiagramme/registrationPage.png)

[search_Sequence](https://github.com/dhbw-ka-tinf22b5-dinder/Dinder-SRS/tree/main/Diagramme/Sequenzdiagramme/search.png)

[swipe_Sequence](https://github.com/dhbw-ka-tinf22b5-dinder/Dinder-SRS/tree/main/Diagramme/Sequenzdiagramme/swipe.png)

### Addition - SOLID:
LSP - We use superclasses and subclasses for the models in the backend. The model interfaces can be replaced by the subclasses (OOP).

**Comment by ghost** [comment]
Kling sehr interessant bei euch. Jedoch habe ich noch kleine Anmerkungen:
- Zum einen hab ich mich gefragt, warum @mausio (der nicht zu eurer Gruppe gehört) einen extra branch und commit bei euch hat (der ziemlich groß ist ~4000 lines in 25 files).
- Dann konnte ich nicht direkt nachvollziehen, was bei der Klasse Advertisement das Attribut **plz** ist. Vielleicht könnt ihr dafür einen explizitieren Namen nehmen?
- Zudem könnt ihr für **charID**, **swipeID**, **MessageID** und **AdvertisementID** anstelle eines **Int** die Klasse **UUID** verwenden. Damit könnt ihr ID´s automatisch und zufällig generieren lassen: **UUID.randomUUID()**.
- Außerdem wäre es schön, wenn ihr in den Klassendiagrammen etwas einheitlicher vorgehen würdet. Zum einen benutzt ihr im Backend Klassendiagramm die Datentypen **Int**, als auch **Integer** oder schreibt manchmal Attribute groß und manchmal klein. Dies gilt auch für den Unterschied zwischen Frontend und Backend (Also Unterschiede zwischen den beiden Klassendiagrammen, wie zB: dass bei dem einen die Datentypen der Attribute stehen und beim anderen nicht).
- Damit ihr nicht, wie ich aus dem Klassendiagramm schließe, die Entities direkt über euren **Controller** an das Frontend schickt und damit womöglich auch Daten weitergebt, die nicht unbedingt nach außen gelangen sollen, könnt ihr stadessen auch für diese **DTO´s** erstellen. Diese können dann über eure **REST API** verteilt werden. Um den **DTO** in die **Entity** und umgekehrt zu transformieren könnt ihr Mapper verwenden.

Zum Abschluss interessiert mich noch warum ihr euch im Frontend für **JavaScript** mit **React** entschieden habt und nicht direkt für **TypeScript** um auch Typen zu haben?

**Reply by mausio** [reply]
> * Zum einen hab ich mich gefragt, warum @mausio (der nicht zu eurer Gruppe gehört) einen extra branch und commit bei euch hat (der ziemlich groß ist ~4000 lines in 25 files).

Servus @babyport,
Ich habe den Gruppenmitglieder*innen von Dinder einen Einblick in React mit Vite, Routing, Redux usw. gegeben und auf ihr bisheriges Front End als Vorlage verwendet.
Das Beispiel habe ich ihnen in einer Branch hochgeladen, um ein "Skelett" zur Verfügung zu stellen :)
